contracts/Stage2/BondStorage.sol

```diff
@@ -114,70 +128,41 @@ contract BondStorage is AccessControl {
-            if (hasExpired(bonds[i]) && !bonds[i].tapped) {
-                tapBond(_user, i); // prevents the abuse of squeezing profit from same bond more than once
-
-                // here we calculate how much we are paying out in sEUR in total and the
-                // profit component, also in sEUR.
-                (uint256 totalPayoutSeuro, uint256 profitSeuro, uint256 totalPayoutOther, uint256 profitOther) = calculateBond(bonds[i]);
-                uint256 payoutTok = seuroToStandardToken(totalPayoutSeuro) + otherTokenToStandardToken(totalPayoutOther);
-                uint256 profitTok = seuroToStandardToken(profitSeuro) + otherTokenToStandardToken(profitOther);
-
-                // increase the user's accumulated profit. only for show or as "fun to know"
-                increaseProfitAmount(_user, profitTok);
-
-                // add the total payout in tokens as a claim. this is the principal in sEURO converted
-                // to TST and the profit in sEUR converted to TST.
-                increaseClaimAmount(_user, payoutTok);
-
-                // one less bond active since this has expired
-                decrementActiveBonds(_user);
-            }
-        }
-    }
-
     function getActiveBonds(address _user) external view returns (uint256) { return issuedBonds[_user].amountBondsActive; }
 
     function getUserBonds(address _user) public view virtual returns (Bond[] memory) { return issuedBonds[_user].bonds; }
 
     function getBondAt(address _user, uint256 index) external view virtual returns (Bond memory) { return getUserBonds(_user)[index]; }
 
-    function getProfit(address _user) external view virtual returns (uint256) { return issuedBonds[_user].profitAmount; }
+    function getProfit(address _user) external view virtual returns (uint256 profitAmount) { 
+        Bond[] memory bonds = getUserBonds(_user);
+        for (uint256 i = 0; i < bonds.length; i++) if (claimable(bonds[i])) profitAmount += bonds[i].profit;
+    }
 
-    function getClaimAmount(address _user) external view virtual returns (uint256) { return issuedBonds[_user].claimAmount; }
+    function getClaimAmount(address _user) public view virtual returns (uint256 claimAmount) {
+        Bond[] memory bonds = getUserBonds(_user);
+        for (uint256 i = 0; i < bonds.length; i++) if (claimable(bonds[i])) claimAmount += bonds[i].reward;
+    }
 
     // Claims the payout in TST tokens by sending it to the user's wallet and resetting the claim to zero.
     function claimReward(address _user) external {
-        uint256 rewardAmount = issuedBonds[_user].claimAmount;
-        require(rewardAmount > 0, "err-no-reward");
-        issuedBonds[_user].claimAmount = 0;
-        tokenGateway.transferReward(_user, rewardAmount);
+        uint256 reward = getClaimAmount(_user);
+        tapUntappedBonds(_user);
+        if (!active(_user)) setInactive(_user);
+        require(reward > 0, "err-no-reward");
```

Comment:
can we move up this check directly after defining this variable? just as a general rule to avoid potentially changing the state in a fail case.

Reply:
good idea. in my head i thought it would be better to mark the bonds are completed, even if the payout is 0, but actually there won't be any bonds with 0 payout so there's no point